crates/functional/src/lib.rs

```diff
@@ -0,0 +1,456 @@
+    props: T::TProps,
+    hook_state: RefCell<Option<HookState>>,
+}
+
+impl<T: 'static> Component for FunctionComponent<T>
+where
+    T: FunctionProvider,
+{
+    type Message = ();
+    type Properties = T::TProps;
+
+    fn create(props: Self::Properties, link: ComponentLink<Self>) -> Self {
+        let render_call = Rc::new(move || link.send_message(()));
+        FunctionComponent {
+            _never: std::marker::PhantomData::default(),
+            props,
+            hook_state: RefCell::new(Some(HookState {
+                counter: 0,
+                render_call,
+                hooks: vec![],
+                next_tick: vec![],
+            })),
+        }
+    }
+
+    fn update(&mut self, _: Self::Message) -> bool {
+        let mut should_rerender = false;
+        for scheduled_fn in self
+            .hook_state
+            .borrow_mut()
+            .deref_mut()
+            .as_mut()
+            .expect("Hook state should be mutable here")
+            .next_tick
+            .drain(..)
+        {
+            should_rerender = scheduled_fn() || should_rerender;
+        }
+        should_rerender
+    }
+
+    fn change(&mut self, props: Self::Properties) -> bool {
+        let mut props = props;
+        std::mem::swap(&mut self.props, &mut props);
+        props == self.props
+    }
+
+    //noinspection DuplicatedCode
+    fn view(&self) -> Html {
+        // Reset hook
+        self.hook_state
+            .try_borrow_mut()
+            .expect("Unexpected concurrent/nested view call")
+            .as_mut()
+            .unwrap()
+            .counter = 0;
+        // Load hook
+        CURRENT_HOOK.with(|previous_hook| {
+            std::mem::swap(
+                previous_hook
+                    .try_borrow_mut()
+                    .expect("Previous hook still borrowed")
+                    .deref_mut(),
+                self.hook_state.borrow_mut().deref_mut(),
+            );
+        });
+
+        let ret = T::run(&self.props);
+
+        // Unload hook
+        let (render_call, has_next_tick) = CURRENT_HOOK.with(|previous_hook| {
+            std::mem::swap(
+                previous_hook
+                    .try_borrow_mut()
+                    .expect("Previous hook still borrowed")
+                    .deref_mut(),
+                self.hook_state.borrow_mut().deref_mut(),
+            );
+            let mut hook_ref_mut = self
+                .hook_state
+                .try_borrow_mut()
+                .expect("Expected hook to be borrowable at this point");
+            let component_hook_state = hook_ref_mut
+                .deref_mut()
+                .as_mut()
+                .expect("Components should always have a hook state");
+            (
+                component_hook_state.render_call.clone(),
+                !component_hook_state.next_tick.is_empty(),
+            )
+        });
+
+        if has_next_tick {
+            render_call();
+        }
```

Comment:
Is this for effects? Effects shouldn't cause renders, this feels like a hack.

Reply:
render_call is a misnomer. It only causes a message to be sent to self, if rerendering happens is decided based on the return values of the next_tick hooks

Reply:
Ah yeah, the name really threw me off.  So I guess there are hooks that can re-render and hooks that won't. I guess I wish the implementation was a bit more explicit about that separation.

Reply:
Renamed